Integrate EIA-923 Annual Environmental Information (Schedule 8C) spreadsheet maps

zaneselvans opened the pull request:
# PR Overview

A draft of spreadhseet mapping metadata for the EIA-923 Schedule 8C, which is about emissions control equipment. Based on @grgmiller's PR #1950.

See notes on issue #2448 

# PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)

grgmiller commented:
@zaneselvans as you work on integrating emission control data, one thing to be aware of is that the raw EIA-923 data contains incomplete `_control_id` mapping. For example, some data rows that contain non-missing so2 control efficiencies have a missing `so2_control_id`. In OGE, we address this by filling in the missing control id with the control id for another pollutant (see: https://github.com/singularity-energy/open-grid-emissions/blob/bb0c0329c54b25cfc0e0c550ebf242bb7930f733/src/emissions.py#L1058) but there may be a better way to address this on the pudl side.

e-belfer commented on src/pudl/package_data/eia923/column_maps/emissions_control.csv:
```diff
@@ -0,0 +1,24 @@
+hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service,hours_in_service
+annual_nox_emission_rate_lb_per_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,nox_emission_rate_entire_year_lbs_mmbtu,n ...
+ozone_season_nox_emission_rate_lb_per_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_emission_rate_may_through_september_lbs_mmbtu,nox_e ...
+pm_emission_rate_lb_per_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu,pm_emissions_rate_lbs_mmbtu
```
Same goes here and for all the records below

e-belfer commented on src/pudl/package_data/eia923/column_maps/emissions_control.csv:
```diff
@@ -0,0 +1,24 @@
+so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date,so2_test_date
+fgd_sorbent_consumption_1000_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons,fgd_sorbent_quantity_thousand_tons
+fgd_electricity_consumption_mwh,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_megawatthours,fgd_electricity_consumption_meg ...
+hg_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency,mercury_removal_efficiency
```
Would be nice to have consistency and use either mercury or hg in our codes. So far, we've been using mercury (e.g. `compliance_year_mercury`).

zaneselvans commented:
Ah yeah you're right I should update all the names. I just took the ones that were here already.

Weirdly, my issue with the 2018 data magically disappeared! I thought it must have been something screwy with my local setup, but if you're getting it now too then I don't know what's going on!

e-belfer commented:
One more weirdness while I'm here! I think the 923 archives aren't a part of the Catalyst organization repo on Zenodo. If I trace the doi for the record I find them here:
https://zenodo.org/record/7682417

But they don't show up in this list: https://zenodo.org/communities/catalyst-cooperative/search?page=1&size=20#

Is there a reason for this, or is this just an oversight? If so, I can make a separate issue.

zaneselvans commented:
No, there's no reason it's not in the community except that adding it to the community has to be done manually, rather than happening automatically when we create a new archive. See  catalyst-cooperative/pudl-archiver#76

I went ahead and added it!

Also updated the naming to match what we did in EIA-860 EnviroEquip.

zaneselvans commented:
Which archive is used is hard-coded in `pudl.workspace.datastore`. Basically every update for every data source breaks the ETL because FERC & EIA are not even attempting to be consistent, so we can only update to new archives by hand.  For instance, when I tried to use the more recent EIA-923 achive, I discovered that they've dropped the FIPS ID column for coalmines in the Fuel Receipts and Costs table (just between the "final" release of 2021 data and the revisions to that "final" data that have been made since October).  There is no way to include data from more than one archive.

One thing that I have been messing around with locally is using Python 3.11. So maybe this problem is magically fixed in 3.11?! Seems unlikely.

I think it would be okay (though definitely not ideal) to merge this in now with the `emissions_control` page commented out in the extraction just to get the metadata into the system and not have it be dangling in a limbo-PR.  I suspect that the next time we update the eia923 DOI it'll work just fine with no weirdness.

e-belfer reviewed (approved):
With the rollback of the commenting out for now, this looks good to merge. I'd suggest leaving a comment in-line both in the assets and in the extraction so nobody's confused about why we have an empty asset, though!

zaneselvans commented:
I've commented out the `emissions_control` asset, but I'm not sure why this would prevent it from attempting to extract the `emissions_control` page. But I also can't make the page fail locally any more so it's hard for me to test what's going on.

Where are the 2 places (assets + extraction) that you're thinking of here?

Maybe I need to add `emissions_control` to the `blacklisted_pages` in the EIA923 excel extractor?

e-belfer commented:
Adding `emissions_control` to the blacklist did the trick. Should be good to go!

zaneselvans commented:
Ah okay great! Thank you! I'll merge once the CI passes. Such a weird issue.
